mdstcpip/mdsipshr/Connections.c

```diff
@@ -465,13 +465,14 @@ static inline int authorize_client(Connection *c, char *username)
 int AddConnection(Connection *c)
 {
   MDSIPTHREADSTATIC_INIT;
-  static int id = INVALID_CONNECTION_ID;
+  
+  // Connection IDs are issued in sequence.
+  // Issue #2625 requires connection IDs starting at 1 to avoid IDL issues,
+  // so initialize this static ID to zero.
+  static int id = INVALID_CONNECTION_ID + 1;
   static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
   pthread_mutex_lock(&lock);
-  do
-  {
-    id++; // find next free id
-  } while (id == INVALID_CONNECTION_ID && _FindConnection(id, NULL, MDSIPTHREADSTATIC_VAR));
```

Comment:
I'm still worried about the unintended side effects of removing this, why was this loop removed again?

Reply:
That is indeed a reasonable worry.   Which is why this change needs scrutiny.   If there are holes in the following argument (and even if not), we can restore the loop.

- Note that `AddConnection()` defines the `id` as a static, thus it maintains its value across calls of this function.   That is an important fact when analyzing the original code.

- The original code, sets `id = INVALID_CONNECTION_ID` (i.e., -1) at compile time.
- And on the first call, the do-while loop increments `id`, so it is now 0.
- Thus the `while (id == INVALID_CONNECTION_ID && ...` will fail on the very first call.
- And it also fails on all subsequent calls (because `id` continually increases)
- And because C uses shortcut evaluation of `&&`, the `_FindConnection()` function is never called
- Thus the entire loop is just an overly complicated way of executing the statement `id++` once per call.
- Meaning that every time a new connection is added, it is just issued the next sequence number.   New connections never re-use an old connection number.
- Which is why the proposed fix tosses the do-while loop and replaces it with `id++`.

One way to determine if there are holes in this argument is to run the code in the debugger.   If it ever calls `_FindConnection()` or if it ever iterates on the loop, then the above reasoning is flawed.

A conjecture (likely wrong) is that years ago, connections were maintained in an array.   In which case, there would need to be a loop to scan through the array to find empty slots (i.e., connections no longer in use that can now be reused.).  Eventually though, the array was replaced by the current scheme which is a linked list of connection structures.   And the looping was moved into the `_FindConnection()` routine -- which is used by more than `AddConnection()`.  The net result is the current code which is just a complicated way to execute `id++`.

However, there is always the risk of an unintended side effect when tinkering with old code.   So if it is safer, we can retain the original do-while loop, and the fix should still work fine.   Because the crux of the fix is the `+ 1` in the following line.

```
static int id = INVALID_CONNECTION_ID + 1;
```

If we do retain the original loop, then we should add a comment explaining that it is just an elaborate way to issue ID numbers sequentially, as that comment will be an aid to future maintenance developers.

Reply:
the point of the loop was rather to allow the id to overrun. and then check if the old id is still in use.. the condition should use a `||` instead of `&&` so it would skip invalid id as well as ids still in use. if you want id 0 to be invalid (i am struggeling with the motivation to do it here) you would need to skip it here as well or redefine the INVALID_CONNECTION_ID.  possobly one would restrict the valid range to positive integers. but as fare are the original implementation is concerned, it is only an id similar to a fileid.

Reply:
Hi @zack-vii -- Thanks for the explanation.   Changing the `&&` to `||` makes a big difference.   In that case, the do-while loop is useful as it provides an additional integrity check on the linked list of connection structures.   

In theory, the static `id` and the linked list should never be out of sync, and thus the proposed fix without the loop should be reliable.

However, it is good defensive programming practice to assume that data structures can be corrupted.   And if that did occur, then it would be a serious problem if the same id was used with multiple structures in the linked list.   Using the do-while loop and `_FindConnection()` is thus a useful integrity check.

The reason that zero is no longer used for the connection id, is because of an oddity of the IDL language.   In IDL, functions and procedures can be given optional "keyword" arguments, such as `socket=socket`.   And there is a related IDL function, `keyword_set()` that can be used to test if a keyword argument is present and has a non-zero value.   Stated another way, when the connection ID is zero, then `keyword_set(socket)` fails, and the IDL code that is in MDSplus doesn't behave as expected.   We evaluated fixes just to the IDL code, but they were complicated and involved a lot of code changes (e.g., slipping in a shim routine everywhere the connection ID is used).   The simplest fix is simply to ensure that connection IDs are integers greater than zero.

I will restore the loop to this PR, but with the `||` change you suggested.

Reply:
Just added the changes that @zack-vii suggested.

Passes a Docker build with the `--test` option.   Also passes all the manual testing shown above in this PR (i.e., connection IDs starting at 1 and issued in sequence).

Please re-review this PR.

Reply:
Hi @zack-vii -- I forgot to explain why `INVALID_CONNECTION_ID` was left as `-1`.   We would indeed like to redefine it as zero.   Unfortunately, a scan of the entire MDSplus source code reveals a few instances where a connection id variable is being compared to the `-1` literal.   So, it was deemed less risky to leave `INVALID_CONNECTION_ID` as is, and just modify `AddConnection()` to skip over zero.